[libc++] Add std::stacktrace (P0881R7)

elsteveogrande opened the pull request:
`<stacktrace>` implementation for C++23.

* Implements https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0881r7.html
* Also implements hardening: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p3697r0.html

## Platform Support

This captures the state of functionality across several platforms and architectures.
Note that no operation is guaranteed if the program isn't compiled with debug information.
In non-Windows platforms where `libunwind.h` or `Unwind.h` are not provided, we cannot capture a trace.

(With the basic requirements of don't-strip-debug-info, and building libcxx on a capable-enough system,
then this matrix _should_ be accurate.)

| Platform | Architecture           | Capture trace | Symbols  | Source Location | Impl notes                        |
|---       |---         |---            |---       |--               |---                                |
| MacOS    | ARM64<br/>x86-64    | ✅            | ⚠️<sup>1</sup> |⚠️<sup>2</sup>     |      |
| Linux / glibc hosted | i686<br/>x86-64 | ✅ | ⚠️<sup>1</sup> |⚠️<sup>2</sup>     | |
| FreeBSD  | i686<br/>x86-64    | ✅ | ⚠️<sup>1</sup> |⚠️<sup>2</sup>     | |
| AIX      | PPC32<br/>PPC64  | ✅ | ⚠️<sup>1</sup>  |⚠️<sup>2</sup>     | |
| Android (NDK 21-23)  | i686 | Ⓧ | Ⓧ      | Ⓧ     | NDK's unwinder is busted in versions 21-23 for i686; traces will be empty |
| Android (other NDKs) | i686<br/>x86-64 | ✅ | ⚠️<sup>1</sup> | ⚠️<sup>2</sup> | |
| Windows  | i686      | Ⓧ<sup>3</sup>          | Ⓧ       |Ⓧ      | Crashes on stacktrace collection; marked unsupported |
| Windows  |  ARM64<br/>x86-64 | ✅            | ✅       |✅      | OS-provided `dbghelp`, `psapi` do everything for us |

<sup>(1)</sup><small>
Symbols not _yet_ available; I'm aiming to better support this soon.
</small><br/>
<sup>(2)</sup><small>
Source file not available; tries to fall back on executable/library name if available.
Line number not available at all.
</small><br/>
<sup>(3)</sup><small>
Collecting the stacktrace fails (process exits with -1).  I marked this platform unsupported without investing more time within this PR.
</small><br/>

Further:

* For those platforms with spotty support, we nevertheless aim to have it compile ok
and not crash at runtime, and in the worst case, hand back an empty stacktrace.
* These operations are not guaranteed to be async-signal-safe.  While that would be ideal --
  especially since a stacktrace might be useful to display during a crash handler --
  that's also not guaranteed in C++23's `stacktrace`.
  - However it _may_ be possible in the future if we carefully figure out how to make that so.
    For example if a user-provided allocator is used (instead of the default `std::allocator`),
    and that allocator is known to be async-signal-safe (perhaps expressly for the reason of
    taking a trace during signal handling), this may be feasible.

Closes #105131
Closes #105167
Closes #104998

elsteveogrande committed:
Add C++23 stacktrace

frederick-vs-ja commented on libcxx/include/__ostream/basic_ostream.h:
```diff
@@ -570,7 +570,7 @@ _LIBCPP_HIDE_FROM_ABI _Stream&& operator<<(_Stream&& __os, const _Tp& __x) {
 
 template <class _CharT, class _Traits, class _Allocator>
-basic_ostream<_CharT, _Traits>&
+_LIBCPP_HIDE_FROM_ABI basic_ostream<_CharT, _Traits>&
```
Why is this `operator<<` touched? IIUC this should belong to another patch.

elsteveogrande replied:
Good catch!  I needed this to make the GCC build work but meant to remove this.

elsteveogrande replied:
Reverted this file

frederick-vs-ja commented on libcxx/include/experimental/stacktrace:
```diff
@@ -0,0 +1,191 @@
+// -*- C++ -*-
```
Thanks!

But while `stacktrace` is now experimental, we should place the headers to the usual place (i.e. under `libcxx/include/` directly). `<experimental/...>` is for features from some technical specializations (TS) that are not standardized yet.

elsteveogrande replied:
Will move.  Thanks @frederick-vs-ja !

Zingam replied:
@elsteveogrande If you want to make this PR "experimental" you can do this via macros. We have a mechanism for experimental library features. I'm sorry I can't give an accurate info right away.

elsteveogrande replied:
Thanks @Zingam @frederick-vs-ja @H-G-Hristov for the feedback!

As for `experimental`: I'll just take all these files out of "experimental" dirs and make this non-experimental.  I just thought new stuff goes into an experimental holding tank until ready to come out :)

If anyone requests it though, I'll add "experimental" flags + macros.

frederick-vs-ja commented:
This PR should have `Closes #105131` `Closes #105167` in the description per [GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

> * Still needs support for `formatter`

IMO this can be done in a later PR...

Zingam commented:
@elsteveogrande Thanks for working on this. It's great to see progress on `<stacktrace>`. I took the liberty to assign the issue to you: https://github.com/llvm/llvm-project/issues/105131

Just a few nits: as @frederick-vs-ja pointed out we have a few unwritten conventions for PRs, which while not required strictly is nice to follow for consistency, such as naming PR, and putting relevant info in the commit message, using the github syntax to link PR to issues.
Please mark also implemented/in  progress papers and LWG issues in the status pages found here: https://github.com/llvm/llvm-project/tree/main/libcxx/docs/Status also add an entry for completed papers to Release Notes in https://github.com/llvm/llvm-project/blob/main/libcxx/docs/ReleaseNotes/21.rst

Zingam commented on llvm/utils/gn/secondary/libcxx/src/BUILD.gn:
```diff
@@ -317,6 +317,40 @@ if (libcxx_enable_experimental) {
     output_dir = runtimes_dir
     output_name = "c++experimental"
     sources = [ "experimental/keep.cpp" ]
+    sources += [
```
I think you shouldn't change this file manually??

Zingam commented on libcxx/test/std/diagnostics/stacktrace/syn.pass.cpp:
```diff
@@ -0,0 +1,147 @@
+  using T8 = decltype(operator<<(*os, std::basic_stacktrace<Alloc>()));
+  static_assert(std::is_same_v<std::ostream&, T8>);
+
+  // [9]
+  // template<> struct formatter<stacktrace_entry>;
+  //   using T9 = std::formatter<std::stacktrace_entry>;
+  //   static_assert(std::is_constructible_v<T9>);
+
+  // TODO(stacktrace23): needs `formatter`
+
+  // [10]
+  // template<class Allocator> struct formatter<basic_stacktrace<Allocator>>;
+  //   using T10 = std::formatter<std::basic_stacktrace<Alloc>>;
+  //   static_assert(std::is_constructible_v<T10>);
+
+  // TODO(stacktrace23): needs `formatter`
```
Formatter is another paper. You should just add what's implemented.

elsteveogrande replied:
I'll leave these TODO's, with a mention of the relevant ticket, #105257

Zingam replied:
We don't usually do that for papers. We track these via status pages and GitHub issues. IMO no need for TODOs but whatever...

Zingam commented on libcxx/src/experimental/stacktrace/osx/osx.h:
```diff
@@ -0,0 +1,31 @@
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP_STACKTRACE_OSX_H
+#define _LIBCPP_STACKTRACE_OSX_H
```
I'm pretty sure that for the macOS you should implement the availability macros: https://github.com/llvm/llvm-project/blob/main/libcxx/include/__configuration/availability.h

elsteveogrande replied:
Thanks again for the feedback @Zingam -- I will revisit this (I don't understand this fully yet :) )

Zingam commented on libcxx/utils/libcxx/header_information.py:
```diff
@@ -170,7 +170,7 @@ def __hash__(self) -> int:
     "rcu",
     "spanstream",
-    "stacktrace",
+    "stacktrace",  # TODO(stacktrace23): remove this when stacktrace is taken out of experimental
```
You should probably just declare this header as implemented even if marked as "experimental" but I am not too sure about it.

Zingam commented on libcxx/modules/std/stacktrace.inc:
```diff
@@ -8,7 +8,9 @@
 export namespace std {
-#if 0
+// TODO(stacktrace23): update this when stacktrace is taken out of experimental
+#if 0 //_LIBCPP_STD_VER >= 23
```
As everywhere else I think this needs to be guarded via the "experimental" macro you should implement.

frederick-vs-ja commented on libcxx/include/experimental/stacktrace:
```diff
@@ -0,0 +1,191 @@
+
+*/
+
+#if __cplusplus < 201103L && defined(_LIBCPP_USE_FROZEN_CXX03_HEADERS)
```
This is meaningless for a C++23 feature.

frederick-vs-ja commented on libcxx/include/experimental/stacktrace:
```diff
@@ -0,0 +1,191 @@
+#else
+#  include <__config>
+
+#  if _LIBCPP_STD_VER >= 23
```
Most of additions in headers should be guarded with `_LIBCPP_STD_VER >= 23`. I _guess_ we can only make `<__config>` included in these files when  `_LIBCPP_STD_VER < 23`.

elsteveogrande replied:
Hi @frederick-vs-ja -- I think that macro typically comes from `<__config>`, so we need that include before the check.  I'm copying this pattern from other recent (C++20, 23) headers like:
https://github.com/llvm/llvm-project/blob/main/libcxx/include/expected#L44

frederick-vs-ja replied:
Yeah. `<__config>` should be unconditionally included in most (if not all) headers of libc++.

frederick-vs-ja commented on libcxx/test/libcxx/stacktrace/simple.o0.nodebug.pass.cpp:
```diff
@@ -0,0 +1,32 @@
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
```
We can use a clearer form now.
```suggestion
// REQUIRES: std-at-least-c++23
```

frederick-vs-ja commented on libcxx/include/experimental/__stacktrace/detail/alloc.h:
```diff
@@ -0,0 +1,103 @@
+  _LIBCPP_HIDE_FROM_ABI std::pmr::list<std::pmr::string> new_string_list() { return new_list<std::pmr::string>(); }
+
+private:
+  std::function<std::byte*(size_t)> __alloc_func_;
+  std::function<void(std::byte*, size_t)> __dealloc_func_;
```
Isn't this too heavy and/or overly type-erasing?

Also, it seems to me that internal use of `pmr` containers will cause the `polymorphic_allocator` to be stored too many times.

elsteveogrande replied:
I was doubting whether I should have been using `pmr` here.  While convenient for me, it might not have been appropriate to apply here.  (I don't see any `pmr` usage inside libcxx anywhere else...)

There is need to do allocations within stacktrace though (other than for `stacktrace_entry` for which the caller may have provided their allocator).  So I think I'll still need alloc/dealloc functions, which delegate to the caller's allocator.  But if there are other better ways I'm happy to consider those :)

frederick-vs-ja replied:
I found that other implementations (MSVC STL & libstdc++) just pass pointers to allocating and deallcating functions and contexts to the separately compiled library, and the allocation strategy is controlled in the header. IMO `pmr` should be avoided (except for the standard-required typedef name `std::pmr::stacktrace`).

Zingam commented on libcxx/src/experimental/stacktrace/osx/osx.h:
```diff
@@ -0,0 +1,31 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP_STACKTRACE_OSX_H
```
```suggestion
#ifndef _LIBCPP_STACKTRACE_MACOS_H
```

Nit: I'd call this macOS.h since macOS isn't OSX for a while and this feature won't be available in any versions called OSX. So I think using "macOS" is more appropriate.

Zingam commented on libcxx/test/std/diagnostics/stacktrace/basic.cmp.pass.cpp:
```diff
@@ -0,0 +1,106 @@
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+#include <stacktrace>
+
+#include <cassert>
+
+/*
+  (19.6.4.4) Comparisons [stacktrace.basic.cmp]
+
+  template<class Allocator2>
+  friend bool operator==(const basic_stacktrace& x,
+                          const basic_stacktrace<Allocator2>& y) noexcept;
+
+  template<class Allocator2>
+  friend strong_ordering operator<=>(const basic_stacktrace& x,
+                                      const basic_stacktrace<Allocator2>& y) noexcept;
+*/
```
I think the current practice is to test a single function, method in a test, e.g.

stacktrace/basic.cmp/equality.pass.pass for `operator==`
stacktrace/basic.cmp/three_way_comparison.pass.cpp for spaceship

you need to put the tests in subdirectories named after the C++ standard sections

Then add the synopsis and then the includes sorted.

```suggestion

// template<class Allocator>
//  class basic_stacktrace

//  template<class Allocator2>
//  friend bool operator==(const basic_stacktrace& x,
//                          const basic_stacktrace<Allocator2>& y) noexcept;


#include <cassert>
#include <stacktrace>
```

Did you have a test for `noexcept`?

elsteveogrande replied:
I'll break these up, reformat, etc.

elsteveogrande replied:
Oh, and also ensure I have something testing noexcept :)

elsteveogrande replied:
I broke up the tests but still need one for `noexcept`

elsteveogrande replied:
Added `noexcept` checks to the unit tests

elsteveogrande commented:
> @elsteveogrande BTW It's better to use "merge" after the review started because rebasing affects the comments on GH.

Argh, yeah I see my re-stacking and force-pushing to my branch messed things up a little; will avoid rebasing

Zingam commented on libcxx/docs/Status/Cxx23Issues.csv:
```diff
@@ -189,7 +189,7 @@
 "`LWG3118 <https://wg21.link/LWG3118>`__","``fpos`` equality comparison unspecified","2022-11 (Kona)","","",""
 "`LWG3177 <https://wg21.link/LWG3177>`__","Limit permission to specialize variable templates to program-defined types","2022-11 (Kona)","|Nothing To Do|","",""
-"`LWG3515 <https://wg21.link/LWG3515>`__","§[stacktrace.basic.nonmem]: ``operator<<`` should be less templatized","2022-11 (Kona)","","",""
+"`LWG3515 <https://wg21.link/LWG3515>`__","§[stacktrace.basic.nonmem]: ``operator<<`` should be less templatized","2022-11 (Kona)","|Nothing To Do|","",""
```
I only had a quick look but I think the status should be `|Complete|` and you also need to add `Closes #104998 `.

https://github.com/llvm/llvm-project/issues/104998 

We don't add an entry to Release Notes for LWG issues.

R-Goc commented on libcxx/src/stacktrace/config.h:
```diff
@@ -0,0 +1,41 @@
+#  define _LIBCPP_STACKTRACE_CAN_SPAWN_TOOLS
+#endif
+
+#if __has_include(<windows.h>) && __has_include(<dbghelp.h>) && __has_include(<psapi.h>)
```
Why not actually check for windows? Either with the usual macros, or ideally a libcxx specific one if it exists.

R-Goc replied:
Same goes for the other platforms

Arthapz commented on libcxx/src/stacktrace/tools.cpp:
```diff
@@ -0,0 +1,437 @@
+  std::string_view tmp;
+  while (true) {
+    std::getline(output, line);
+    while (isspace(line.back())) {
```
tryed this PR on my linux, and crashed here
fixed the crash by changing the line to 
```cpp
while (!line.empty() && isspace(line.back())) {
```

elsteveogrande replied:
Nice, thanks for finding this @Arthapz !

elsteveogrande commented on libcxx/include/__stacktrace/basic.h:
```diff
@@ -0,0 +1,237 @@
+  current(size_type __skip,
+          size_type __max_depth,
+          const allocator_type& __caller_alloc = allocator_type()) noexcept(__kNoThrowAlloc) {
+    __stacktrace::base __builder(__caller_alloc);
```
Thanks @var-const !  I'm 99% sure the implementation is ok but I'll ensure overflow isn't possible.

For this I will do the following (please confirm):

* `_LIBCPP_ASSERT_`*something* to check this within `current`
* add an `XFAIL` test for this, with e.g. `skip=1` and `max_depth=0xffff_ffff_ffff_ffff`
* set `__cpp_lib_hardened_basic_stacktrace` to `202502`

elsteveogrande replied:
Actually I'm realizing I set the default max_depth to `~0`, i.e. max unsigned value, which is sure to trigger that assertion :)  Will fix that

elsteveogrande commented on libcxx/include/string:
```diff
@@ -3966,19 +3966,19 @@ _LIBCPP_HIDE_FROM_ABI basic_istream<_CharT, _Traits>&
-inline _LIBCPP_HIDE_FROM_ABI basic_istream<_CharT, _Traits>&
+inline _LIBCPP_EXPORTED_FROM_ABI basic_istream<_CharT, _Traits>&
 getline(basic_istream<_CharT, _Traits>&& __is, basic_string<_CharT, _Traits, _Allocator>& __str);
 
```
Question: should these be `EXPORTED_FROM_ABI`?  I couldn't access these definitions without exporting them.  Not sure why they were hidden, or how else to solve this.

The signatures appear to match the public API.  [cppreference link, I didn't locate the paper these were introduced in](https://en.cppreference.com/w/cpp/string/basic_string/getline.html)

(If these should actually stay `EXPORTED_` then I can add this as a separate patch, along with any other unrelated changes)

elsteveogrande replied:
Ignore, will undo this

elsteveogrande commented on libcxx/src/stacktrace/fd.h:
I added a couple of "utility" or "helper" files like this.  If there are better ways to do this, e.g. other libcxx-internal files containing file-descriptor-for-reading-process-stdout, I'll be happy to use those instead, to reduce this huge PR

elsteveogrande reviewed (commented):
Still in draft mode, but I think we're in the home stretch :)

General comments:
* This PR is huge, sorry.  I could break this up and ship in smaller pieces instead.
* I'm watching CI and whack-a-moleing problems as they appear
* It's been a bit difficult to ensure this "works everywhere".  Depending on the target system -- which can mean CI, the build machine, or the target on which this ultimately runs -- the available headers and libraries can differ.  I guess that's part of why this is huge.  Let me know if I'm way off-track with how I'm going about this

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+  _LIBCPP_ALWAYS_INLINE static basic_stacktrace
+  current(const allocator_type& __caller_alloc = allocator_type()) noexcept(__kNoThrowAlloc) {
+    size_type __skip      = 0;
+    size_type __max_depth = __default_max_depth;
```
Why is this a local variable?

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+    return __ret;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr ~basic_stacktrace() = default;
```
Why is this `constexpr`?

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+      : basic_stacktrace(std::move(__other), __other.__alloc_) {}
+
+  _LIBCPP_HIDE_FROM_ABI basic_stacktrace& operator=(const basic_stacktrace& __other) {
+    if (std::addressof(__other) != this) {
+      if (__kPropOnCopyAssign) {
+        new (this) basic_stacktrace(__other, __other.__alloc_);
+      } else {
+        new (this) basic_stacktrace(__other);
+      }
+    }
+    return *this;
```
Why do you construct a new `basic_stacktrace`? This leaks memory. Please add a regression test.

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+  // (19.6.4.3)
+  // [stacktrace.basic.obs], observers
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI allocator_type get_allocator() const noexcept { return __alloc_; }
```
Please add tests for all `[[nodiscard]]` applications.

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+// Non-member functions [stacktrace.basic.nonmem]
+
+template <class _Allocator>
+_LIBCPP_HIDE_FROM_ABI inline void
```
```suggestion
_LIBCPP_HIDE_FROM_ABI void
```
Throughout.

philnik777 commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,342 @@
+template <class _Allocator>
+struct hash<basic_stacktrace<_Allocator>> {
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI size_t operator()(basic_stacktrace<_Allocator> const& __context) const noexcept {
+    size_t __ret = 1;
+    for (auto const& __entry : __context.__entries_) {
+      __ret += hash<uintptr_t>()(__entry.native_handle());
+    }
+    return __ret;
+  }
```
This seems like a rather poor hash functions. We should probably just call `__hash_memory` on the underlying vector.

philnik777 commented on libcxx/include/__stacktrace/stacktrace_entry.h:
```diff
@@ -0,0 +1,216 @@
+  bool operator==(str_alloc<_Tp> const& __rhs) const { return std::addressof(__rhs) == this; }
+};
+
+struct str : basic_string<char, char_traits<char>, str_alloc<char>> {
+  using base = basic_string<char, char_traits<char>, str_alloc<char>>;
+  _LIBCPP_HIDE_FROM_ABI str(str_alloc<char> const& __alloc) : base(__alloc) {}
+  _LIBCPP_HIDE_FROM_ABI string_view view() const { return {this->data(), this->size()}; }
+};
```
Why?

philnik777 commented on libcxx/include/module.modulemap.in:
```diff
@@ -2010,6 +2010,31 @@ module std [system] {
+    header "stacktrace"
+    export *
+    // TODO: Workaround for https://github.com/llvm/llvm-project/issues/120108
+    export std.cstddef.byte
+    export std.cstddef.size_t
+    export std.format.escaped_output_table
+    export std.format.formatter
+    export std.format.width_estimation_table
+    export std.functional.function
+    export std.functional.hash
+    export std.iterator.iterator
+    export std.iterator.reverse_iterator
+    export std.memory.allocator
+    export std.memory.allocator_traits
+    export std.memory_resource.polymorphic_allocator
+    export std.new.allocate
+    export std.ostream.basic_ostream
+    export std.type_traits.is_nothrow_constructible
+    export std.vector.vector
```
How are these so many? This isn't usually required, and some of them aren't required to be available from `<stacktrace>` at all AFAICT.

philnik777 commented on libcxx/include/stacktrace:
```diff
@@ -0,0 +1,187 @@
+#  if _LIBCPP_STD_VER >= 23
+#    include <__stacktrace/basic_stacktrace.h>
+#    include <__stacktrace/stacktrace_entry.h>
+#    include <compare> // per [stacktrace.syn]
```
Let's include this unconditionally. We don't include it conditionally anywhere else AFAIK.

philnik777 commented on libcxx/include/streambuf:
```diff
@@ -420,7 +420,7 @@ private:
 
   template <class _CharT2, class _Traits2, class _Allocator>
-  _LIBCPP_HIDE_FROM_ABI friend basic_istream<_CharT2, _Traits2>&
+  _LIBCPP_EXPORTED_FROM_ABI friend basic_istream<_CharT2, _Traits2>&
```
This is both unrelated and incorrect.

philnik777 commented on libcxx/lib/abi/arm64-apple-darwin.libcxxabi.v1.stable.exceptions.nonew.abilist:
```diff
@@ -1305,7 +1314,6 @@
 {'is_defined': True, 'name': '__ZNSt3__113shared_futureIvED1Ev', 'type': 'FUNC'}
 {'is_defined': True, 'name': '__ZNSt3__113shared_futureIvED2Ev', 'type': 'FUNC'}
 {'is_defined': True, 'name': '__ZNSt3__113shared_futureIvEaSERKS1_', 'type': 'FUNC'}
-{'is_defined': True, 'name': '__ZNSt3__113__hash_memoryEPKvm', 'type': 'FUNC'}
```
Please remove the unrelated changes here and elsewhere.

philnik777 commented on libcxx/src/CMakeLists.txt:
```diff
@@ -335,6 +340,7 @@ endif()
 
 add_library(cxx_experimental STATIC ${LIBCXX_EXPERIMENTAL_SOURCES})
 target_link_libraries(cxx_experimental PUBLIC cxx-headers)
+
```
Again, unrelated change.

philnik777 commented on libcxx/src/stacktrace/images.cpp:
```diff
@@ -0,0 +1,92 @@
+
+#include "__config"
+
+#if defined(__APPLE__)
```
Please split these up into separate headers.

philnik777 commented on libcxx/src/stacktrace/images.h:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  /** Image representing the main program, or nullptr if we couldn't find it */
+  _LIBCPP_HIDE_FROM_ABI image* main_prog_image() {
```
You don't need `_LIBCPP_HIDE_FROM_ABI` in the dylib.

philnik777 commented on libcxx/src/stacktrace/impl.cpp:
```diff
@@ -0,0 +1,84 @@
+  // Although 64-bit addresses are 16 nibbles long, they're often <= 0x7fff_ffff_ffff
+  constexpr static int __k_addr_width = (sizeof(void*) > 4) ? 12 : 8;
+
+  __os << "0x" << std::hex << std::setfill('0') << std::setw(__k_addr_width) << __addr_;
```
You can't just change the state of the stream like this. Please add a regression test.

elsteveogrande replied:
Ah!  Good call, will avoid this.

I was thinking, maybe I could implement a `formatter` for these -- although perhaps in this PR, not including the [formatting options as prescribed here](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2693r1.pdf) , and have `operator<<` go through that instead?

philnik777 commented on libcxx/src/stacktrace/impl_generic.cpp:
```diff
@@ -0,0 +1,91 @@
+
+_LIBCPP_NO_TAIL_CALLS _LIBCPP_NOINLINE _LIBCPP_EXPORTED_FROM_ABI void
+base::current_impl(size_t skip, size_t max_depth) {
+  if (!max_depth) [[unlikely]] {
```
Please don't add annotations like this without a clear performance benefit.

philnik777 commented on libcxx/src/stacktrace/impl_generic.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+void base::find_symbols() {
+  // TODO
```
We shouldn't introduce these if we don't use them.

philnik777 commented on libcxx/src/stacktrace/impl_windows.cpp:
```diff
@@ -0,0 +1,248 @@
+
+// clang-format off
+
+struct dbghelp_dll final : dll {
```
What doesthis do?

philnik777 commented on libcxx/src/stacktrace/unwinding.h:
```diff
@@ -0,0 +1,108 @@
+      break;
+    }
+    auto& entry = base.__entry_append_();
+    auto& eb    = (__stacktrace::entry_base&)entry;
```
Please avoid the C-style casts all over the place.

philnik777 commented on libcxx/src/stacktrace/unwinding.h:
```diff
@@ -0,0 +1,108 @@
+_LIBCPP_END_NAMESPACE_STD
+
+#  elif __has_include(<unwind.h>)
+#    include <unwind.h>
```
Doesn't libunwind also have `<unwind.h>`? Do we need the separate implementations?

philnik777 commented on libcxx/test/libcxx/stacktrace/only_uses_allocator.pass.cpp:
```diff
@@ -0,0 +1,112 @@
+#include <memory>
+#include <stacktrace>
+
+/*
```
Please don't use C-style comments. Also, please move the comment above the includes.

philnik777 commented on libcxx/test/libcxx/stacktrace/only_uses_allocator.pass.cpp:
```diff
@@ -0,0 +1,112 @@
+#include <stacktrace>
+
+/*
+ * This file includes tests which ensure any allocations performed by `basic_stacktrace`
+ * are done via the user-provided allocator.  We intercept the usual ways to allocate,
+ * counting the number of calls, through and not through the allocator.
+ * (This won't work properly with sanitizers, hence the `UNSUPPORTED` above.)
```
We already have utilities to test this. Why aren't they used here?

elsteveogrande commented:
Ah!  Thanks again, that aligns with the comment from @Zingam a while back, that I didn't quite understand, haha
https://github.com/llvm/llvm-project/pull/136528#pullrequestreview-2780692982

I'm still not familiar with availability but I'll research this (and look at other features)

elsteveogrande reviewed (commented):
Thanks all!  Updating and working through the list :)

Zingam commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -41,6 +41,8 @@ Implemented Papers
 - P2321R2: ``zip`` (`Github <https://github.com/llvm/llvm-project/issues/105169>`__) (The paper is partially
   implemented. ``zip_transform_view`` is implemented in this release)
 - P3168R2: Give ``std::optional`` Range Support (`Github <https://github.com/llvm/llvm-project/issues/105430>`__)
+- P0881R7: A Proposal to add stacktrace library (`Github <https://github.com/llvm/llvm-project/issues/105131>`__)
+- P2301R1: Add a `pmr` alias for `std::stacktrace` (`Github <https://github.com/llvm/llvm-project/issues/105167>`__)
```
```suggestion
- P0881R7: A Proposal to add stacktrace library (`Github <https://llvm.org/PR105131>`__)
- P2301R1: Add a `pmr` alias for `std::stacktrace` (`Github <https://llvm.org/PR105167>`__)
```

Please update all issue links to the universal format after: https://github.com/llvm/llvm-project/pull/156288

elsteveogrande committed:
Formatting, fix generated files, etc.

elsteveogrande committed:
Support Windows 32-bit (and other smaller fixes)

elsteveogrande committed:
Update ABI list, WTF edition.  Hopefully fixed

elsteveogrande committed:
De-uglify issue links

elsteveogrande commented on libcxx/include/__configuration/availability.h:
```diff
@@ -291,6 +291,10 @@
 
+// This controls the availability of C++23 <stacktrace>.
+#define _LIBCPP_AVAILABILITY_HAS_STACKTRACE _LIBCPP_INTRODUCED_IN_LLVM_21
+#define _LIBCPP_AVAILABILITY_STACKTRACE _LIBCPP_INTRODUCED_IN_LLVM_21_ATTRIBUTE
```
@Zingam I think I've used this correctly, please review :)

Sorry, I rebased and force-pushed (the rebase of many diffs was getting too difficult in this long-running PR) and comments got detached.  [Original here](https://github.com/llvm/llvm-project/pull/136528#issuecomment-3243758564)

elsteveogrande committed:
Impl stack-height-sensitive code in headers, remove some macros

elsteveogrande committed:
_Ugly_Names

elsteveogrande commented on libcxx/include/__config:
```diff
@@ -965,6 +965,33 @@ typedef __char32_t char32_t;
+// call-stack frames.
+// clang-format off
+
+// Disables tail-call optimization for "outbound" calls
```
Moved all stack-sensitive (unwinding) code into headers and gave then always-inline, so I'm removing these

elsteveogrande committed:
Split up impl's to make more readable.  Simplify things considerably

elsteveogrande committed:
Clean up windows impl

elsteveogrande committed:
More formatting and generated-file fixes

elsteveogrande committed:
Try to fix windows mingw build

elsteveogrande committed:
Try fixing windows again

elsteveogrande committed:
Update headers to match standard text; fix windows impl; fix abilist

elsteveogrande committed:
More fixes related to standard conformance

elsteveogrande committed:
Improved hash funcs

elsteveogrande committed:
Hashing etc.

elsteveogrande committed:
StringWrapper, Windows

elsteveogrande committed:
Clean up tests and fix some 'noexcepts'

elsteveogrande committed:
ABI lists

elsteveogrande commented:
Thanks for reviewing @jwakely !  I will adjust these names per convention, and as you mentioned (as @philnik777 did in previous drafts that I unfortunately rebased over) I also shouldn't mess with the output stream.

Somewhat frustratingly I've been running into CI errors related to ABI symbols and even unrelated things like which headers are included by other headers (which I didn't touch in this PR).  And even more frustratingly I rebased and force-pushed because I kept having to deal with conflicts in my stack -- but that wiped out comments.  I addressed most of those but will note those areas which I still need to address.

Anyway, I'm still working on this, and I think almost done with the first round of changes.  Later this can be improved with symbol and line number resolution...

elsteveogrande committed:
Fix ABI... ?

elsteveogrande commented on libcxx/test/std/diagnostics/stacktrace/basic.obs/crbegin_crend.pass.cpp:
```diff
@@ -0,0 +1,45 @@
+#include <stacktrace>
+
+// Call chain is: main -> c -> b -> a -> stacktrace::current
+_LIBCPP_NOINLINE std::stacktrace a() { return std::stacktrace::current(); }
```
TODO: use TEST_NO_INLINE instead

elsteveogrande commented on libcxx/include/module.modulemap.in:
```diff
@@ -2010,6 +2010,31 @@ module std [system] {
+
+    header "stacktrace"
+    export *
+    // TODO: Workaround for https://github.com/llvm/llvm-project/issues/120108
```
Let me try cleaning this up, and ensuring this works without so many exports

elsteveogrande commented on libcxx/src/CMakeLists.txt:
```diff
@@ -248,6 +254,11 @@ if (LIBCXX_ENABLE_SHARED)
+  string(APPEND link_libraries " ${CMAKE_LINK_LIBRARY_FLAG}dbghelp")
+  string(APPEND link_libraries " ${CMAKE_LINK_LIBRARY_FLAG}psapi")
+endif()
+
```
https://github.com/llvm/llvm-project/actions/runs/19397040621/job/55499528819?pr=136528

elsteveogrande committed:
Fix includes on a few tests

elsteveogrande committed:
Fix macros in tests

elsteveogrande committed:
Avoid iomanip for operator<<

elsteveogrande committed:
Try a fix for windows impl

elsteveogrande committed:
Fix tests

elsteveogrande committed:
Fix windows pragma comments

elsteveogrande committed:
Windows: undo cmakefile changes, use dll

elsteveogrande committed:
Fix windows

elsteveogrande committed:
Fix names

elsteveogrande committed:
Clean up public headers a little

elsteveogrande committed:
Remove unneeded HIDE_FROM_ABI, other small changes

elsteveogrande committed:
mingw 32-bit fix

elsteveogrande committed:
Update libcxx/include/__stacktrace/basic_stacktrace.h

Co-authored-by: Hristo Hristov <hghristov.rmm@gmail.com>

elsteveogrande committed:
Update libcxx/include/stacktrace

Co-authored-by: A. Jiang <de34@live.cn>

elsteveogrande commented:
Thanks @H-G-Hristov and @frederick-vs-ja , I added those changes, and will need to:
1. Add `nodiscard` tests
2. Fix conflicts
3. And might have to punt `mingw-i686` for now.  I haven't dealt with windows in a very long time, and 32-bit windows in a local VM and in CI have been giving me the same mystery crashes.  I'm 70% sure it's not my code causing the crash.

elsteveogrande committed:
Merge branch 'main' into stacktrace23

R-Goc commented on libcxx/src/stacktrace/windows_impl.cpp:
```diff
@@ -0,0 +1,295 @@
+}
+
+// clang-format off
+bool(WINAPI* EnumProcessModules)(HANDLE, HMODULE*, DWORD, DWORD*);
```
These are not defined as bool, but BOOL. A BOOL is not a bool, but an int under the hood. This applies to all the functions here from what I could see. In other parts of windows api some might return BOOLEAN(which is a BYTE actually).

elsteveogrande replied:
ah, could explain why I had crashes when I tested this on windows way back when :)

frederick-vs-ja commented on libcxx/src/stacktrace/windows_impl.cpp:
```diff
@@ -0,0 +1,295 @@
+    // be overwritten.
+    IMAGEHLP_MODULE mod_info;
+    memset(&mod_info, 0, sizeof(mod_info));
+    mod_info.SizeOfStruct = sizeof(mod_info);
```
Ditto, `{}` or designated initialization.

frederick-vs-ja commented on libcxx/src/stacktrace/images.h:
```diff
@@ -0,0 +1,115 @@
+ */
+struct _Images {
+  constexpr static size_t k_max_images = 256;
+  std::array<_Image, k_max_images + 2> images_{}; // space for the L/R sentinels
```
Can we use a built-in array here?

frederick-vs-ja commented on libcxx/src/stacktrace/images.h:
```diff
@@ -0,0 +1,115 @@
+  /** Get prog_image by index (0 <= index < count_) */
+  _Image& operator[](size_t __index) {
+    _LIBCPP_ASSERT(__index < count_, "index out of range");
+    return images_.at(__index);
```
Is it really intended that the index is always checked, and an exception is thrown when libc++'s assertion is disabled?

elsteveogrande replied:
Good catch, let me fix this (and use `_Image[]` instead of `std::array<...>`)

R-Goc commented on libcxx/src/stacktrace/windows_impl.cpp:
```diff
@@ -0,0 +1,295 @@
+
+  attempted = true;
+
+  HMODULE psapi   = LoadLibraryA("psapi.dll");
```
We want to use LoadLibraryExW(L"dllname.dll", NULL, LOAD_LIBRARY_SEARCH_SYSTEM32) to prevent dll hijacking.

R-Goc commented on libcxx/src/stacktrace/windows_impl.cpp:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  char exe_dir[MAX_PATH];
+  if (!GetModuleFileNameA(nullptr, exe_dir, sizeof(exe_dir))) {
```
GetModuleFileName returns the number of characters written so we should check that instead of using strlen. 

You are not handling the case where the path is truncated due to insufficient buffer length when the size written is equal to the length of the string and GetLastError returns ERROR_INSUFFICIENT_BUFFER. This would happen with long paths. The Ansi version of the API is not long path safe. As a matter of fact what you will get is the beginning that then gets cut off. For example for the path that is ```\\?\E:\prog\test\AVeryLongFDirectoryNameThatHas254CharactersToTestWhetherOrNotLongPathsAreProperlySupportedInYourApplicationToDoSoUseWideVersionsOfTheApisAndTheLongPathPrefixWhereAppropriateOrEnableTheRegistyKeyAndAddLongPathAwareToTheAppManifestThisIsToMakeTheResultLong\TestExecutable.exe```
What GetModuleFileNameA returns is:
```\\?\E:\prog\test\AVeryLongFDirectoryNameThatHas254CharactersToTestWhetherOrNotLongPathsAreProperlySupportedInYourApplicationToDoSoUseWideVersionsOfTheApisAndTheLongPathPrefixWhereAppropriateOrEnableTheRegistyKeyAndAddLongPathAwareToTheAppManifestThisIsToMakeT```
You need to use the wide version of the api and handle long paths which will have the `\\?\` prefix. 

Testing this locally is a bit of a pain as both cmd and powershell fail on long paths badly along with many other programs, so you need to move a program into a long directory with another program you write, and also run it with CreateProcessW, as otherwise what will fail is CMD or Powershell.

elsteveogrande committed:
Merge branch 'main' into stacktrace23

jakeegan committed:
Implement std::stracktrace for AIX

elsteveogrande committed:
Merge branch 'main' into stacktrace23

H-G-Hristov commented on libcxx/docs/Status/Cxx23Issues.csv:
```diff
@@ -189,7 +189,7 @@
 "`LWG3118 <https://wg21.link/LWG3118>`__","``fpos`` equality comparison unspecified","2022-11 (Kona)","","","`#104996 <https://github.com/llvm/llvm-project/issues/104996>`__",""
 "`LWG3177 <https://wg21.link/LWG3177>`__","Limit permission to specialize variable templates to program-defined types","2022-11 (Kona)","|Nothing To Do|","","`#104997 <https://github.com/llvm/llvm-project/issues/104997>`__",""
-"`LWG3515 <https://wg21.link/LWG3515>`__","§[stacktrace.basic.nonmem]: ``operator<<`` should be less templatized","2022-11 (Kona)","","","`#104998 <https://github.com/llvm/llvm-project/issues/104998>`__",""
+"`LWG3515 <https://wg21.link/LWG3515>`__","§[stacktrace.basic.nonmem]: ``operator<<`` should be less templatized","2022-11 (Kona)","|Complete|","","`#104998 <https://github.com/llvm/llvm-project/issues/104998>`__",""
```
```suggestion
"`LWG3515 <https://wg21.link/LWG3515>`__","§[stacktrace.basic.nonmem]: ``operator<<`` should be less templatized","2022-11 (Kona)","|Complete|","23","`#104998 <https://github.com/llvm/llvm-project/issues/104998>`__",""
```

Also the remaining status pages should be updated.

elsteveogrande committed:
Merge branch 'llvm:main' into stacktrace23

elsteveogrande committed:
Updates

elsteveogrande committed:
Fix libcxx/src/stacktrace/images.h, don't use std::array

elsteveogrande committed:
Update windows impl, use BOOL and LPDWORD and other windows-isms

elsteveogrande commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,360 @@
+  // Non-windows: impl separated out into several smaller platform-dependent parts.
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_ALWAYS_INLINE void populate_addrs(size_t __skip, size_t __depth);
+  _LIBCPP_EXPORTED_FROM_ABI void populate_images();
+#  endif
```
* `hash_code` it is
* `base` -> `__base`
* others are now uglified

elsteveogrande committed:
Fix some names

elsteveogrande committed:
Merge pull request #1 from jakeegan/stacktrace23-aix

Implement std::stracktrace for AIX

oold commented on libcxx/src/stacktrace/windows_impl.cpp:
```diff
@@ -0,0 +1,290 @@
+  if (!strstr(sym_path, exe_dir)) {
+    (void)strncat(sym_path, ";", sizeof(sym_path) - 1);
+    (void)strncat(sym_path, exe_dir, sizeof(sym_path) - 1);
+    if (!SymSetSearchPath(proc, sym_path)) {
```
Call [`SymSetSearchPathW`](https://learn.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-symsetsearchpathw) here and use `SymGetSearchPathW` and `GetModuleFileNameW` above to avoid code page issues and support long paths.

elsteveogrande replied:
Thank you @oold and @R-Goc for help on the windows front!  I'm going to adapt [these helper functions in `filesystem/path.cpp`](https://github.com/elsteveogrande/llvm-project/blob/main/libcxx/src/filesystem/path.cpp#L422) or mainly just use `WideCharToMultiByte`

elsteveogrande committed:
Fixes...

elsteveogrande committed:
Fixing windows impl...

elsteveogrande committed:
Use LoadLibraryExW instead of LoadLibraryA

elsteveogrande committed:
Ugly names

elsteveogrande committed:
Uglify more names

elsteveogrande committed:
Fix iostream usage

elsteveogrande committed:
Use GetModuleHandleW (not -A)

elsteveogrande committed:
fix

elsteveogrande committed:
fix

elsteveogrande committed:
Fix for picolibc (has no dlfcn.h)

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
fix

elsteveogrande committed:
fix?

elsteveogrande committed:
fix?

elsteveogrande committed:
fix?

elsteveogrande committed:
Fix for picolibc

elsteveogrande committed:
ABILists

elsteveogrande committed:
ABILists

elsteveogrande committed:
More abilist, csv (transitive include) files

elsteveogrande committed:
fix

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
apple-configuration abi list

elsteveogrande committed:
Adding cout prints to tests

elsteveogrande committed:
Fix abilist

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Merge branch 'main' into stacktrace23

Zingam commented on libcxx/include/__stacktrace/basic_stacktrace.h:
```diff
@@ -0,0 +1,365 @@
+  // (19.6.4.3)
+  // [stacktrace.basic.obs], observers
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI allocator_type get_allocator() const noexcept {
```
You have marked some functions as `[[nodiscard]]` but I don't see the test. Please make sure to apply `[[nodiscard]]` everywhere and also implement the test. See 
- https://github.com/llvm/llvm-project/issues/172124
- https://libcxx.llvm.org/TestingLibcxx.html

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Fix availability.py

elsteveogrande committed:
Update to LLVM 24 + add a missing test

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
fix

elsteveogrande committed:
Fix transitive includes

elsteveogrande committed:
Blindly include compare as per the standard (even if C++03 mode, in which neither stacktrace nor compare were things)

elsteveogrande committed:
Regenerate

elsteveogrande committed:
Regenerate

elsteveogrande committed:
Remove std::cout during unit tests, unbreak no-localization test

elsteveogrande committed:
Merge branch 'llvm:main' into stacktrace23

elsteveogrande committed:
Add _LIBCPP_INTRODUCED_IN_LLVM_24

elsteveogrande committed:
Correct wide-string function usage

elsteveogrande committed:
Fix unwind dependency problem for libc builds

elsteveogrande committed:
Fix test infra for arm (unwind tables)

elsteveogrande committed:
Avoid tail/inline calls, ensure predictable call seq

elsteveogrande committed:
Fix for armv7 tests

elsteveogrande committed:
Fix more ARM builds; add an 'unsupported' test, ensure no crash if stacktrace symbolizing known not to work

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Tidy things up a bit; fix more builds

elsteveogrande committed:
Update abilists

elsteveogrande committed:
Update abilists

elsteveogrande committed:
Minor

elsteveogrande committed:
More abilists

elsteveogrande committed:
Fix arm builds again

elsteveogrande committed:
Fix for windows clang-cl tests

elsteveogrande committed:
On windows, use the better-working CaptureStackBackTrace

elsteveogrande committed:
Not working on windows i686; revert last change which didn't help (and broke more); un-support windows i686

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Merge branch 'main' into stacktrace23

elsteveogrande committed:
Merge branch 'main' into stacktrace23

oold commented:
The current Windows solution uses dbghelp to symbolize the trace. If any other thread uses it too, it might cause UB. AFAIK, Microsoft's STL uses the DIA SDK to symbolize out-of-process, which avoids thread safety issues. Should that be used instead?
